### ISSUE TITLE
Can you tell me how I can use BlendedAkkaActivator?

### ISSUE THREAD

**Issue by PhilAndrew** [woq-blended/blended#61]
Hi, can you give me a small example or tell me how I can use BlendedAkkaActivator in place of DominoActivator?

**Comment by atooni** [comment]
Hi, 

thanks for asking and sorry we are lagging behind with the documentation. We are currently working towards the 2.0 release to clean this up. Meanwhile here are some hints:

There are some examples in blended itself, 
* [blended spray example](https://github.com/woq-blended/blended/blob/master/blended-samples/blended-samples-spray-helloworld/src/main/scala/blended/samples/spray/helloworld/internal/initializers.scala)
* [blended persistence](https://github.com/woq-blended/blended/blob/master/blended-persistence/src/main/scala/blended/persistence/internal/PersistenceActivator.scala)

You still extend from DominoActivator and simply mix in the trait ActorSystemWatching. The cfg parameter is an instance of an [OSGIActorConfig](https://github.com/woq-blended/blended/blob/master/blended-akka/src/main/scala/blended/akka/OSGIActorConfig.scala). This config encapsulates references to the Actor System, an instance of a [ContainerIdentifierService](https://github.com/woq-blended/blended/blob/master/blended-container-context/src/main/scala/blended/container/context/ContainerIdentifierService.scala) and the actor config.

Within the execution block you have to call *setupBundleActor* with the instance of the OSGIActorConfig and a properly constructed Props. *setupBundleActor* will then create a capsule to manage the kifecycle of the bundle actor. The actor will be created upon capsule start and will be available within the Actor System as /user/${bundleSymbolicName}. Upon capsule stop the actor will be stopped as well. 

The config needs a bit more elaboration: There is only one Actor system within a blended container, which reads its config as normal from the application.conf in the Container Config Directory exposed by the ContainerIdentifierService via a ContainerContext. 

Normally, an actor needs additional config and we had to provide a way for actors to contribute their own config. The ContainerIdentifierService/ContainerContext provides a config locator which loads either a file named ${bundleSymbilicName}.conf from the container config directory or as a fallback the section from the application.conf that has ${bundleSymbolicName} as the config key. In both cases, the config object that is passed within the OSGIActorConfig has only(!!!!) the settings of that particular config block and does not inherit anything from application.conf. 

However, when you need global configuration settings, you can use helper methods within [OSGIActor](https://github.com/woq-blended/blended/blob/master/blended-akka/src/main/scala/blended/akka/OSGIActor.scala). The method *bundleActorConfig* combines the values from the isolated actor config described above with an existing application conf. Within that combined config the actor specific block is again available at ${bundleSymbolicName}. 

We found that this way of configuring bundle actors allows each bundle to provide its own configuration without trying to modify the application.conf using class loader magic. Everything is neatly provided via  collaborating OSGI services. 


To summarize:

You need the Actor System exposed as a service and also the a Container Identifier Service. You can have a look at the *blended-karaf-demo* or the *blended-launcher-demo* projects for a complete container packaging, but an Actor System and a Container Identifier Service is the minimum required. 

You should inherit your Actor from OSGIActor to leverage our config pattern. 

You can use a DominoActivator with ActorSystemWatching mixed in as shown in the examples.


I hope this helps
Andreas

**Comment by PhilAndrew** [comment]
Thanks for all the information, I'm trying to understand now what triggers whenActorSystemAvailable to occur, what is the code which causes this to happen, can you tell me?

    def whenActorSystemAvailable(f: OSGIActorConfig => Unit) = {
      val m = new ActorSystemCapsule(capsuleContext, f, bundleContext)
      capsuleContext.addCapsule(m)
    }

Is it the startup of BlendedAkkaActivator ?

I see at this line

      log info "Registering Actor System as Service."
      registerService(osgiContext, system)

Is this code what triggers the above?

**Comment by PhilAndrew** [comment]
What I have done is taken your code related to Akka and split it into 3 projects using Sbt using the https://github.com/wav/osgi-tooling. So far its ok up until whenActorSystemAvailable which it not yet occuring. I'll keep investigating why.

**Comment by atooni** [comment]
HI again, 

to understand what triggers *whenActorSystemAvailable*, you can have a look at [ActorSystemCapsule](https://github.com/woq-blended/blended/blob/master/blended-akka/src/main/scala/blended/akka/internal/ActorSystemCapsule.scala). [ActorSystemWatching](https://github.com/woq-blended/blended/blob/master/blended-akka/src/main/scala/blended/akka/ActorSystemWatching.scala) is merely some Domino-like syntactic sugar. 

There are two dependencies which will trigger *whenActorSystemAvailable*. 

* The Actor system exposed as a service. The bundle [blended-akka](https://github.com/woq-blended/blended/tree/master/blended-akka) provides an activator which starts the Actor system and also provides some of the convenience API to use Scala and Actors in other bundles. All you have to do is to include this bundle in your OSGI setup. The bundle requires some dependencies like Scala and Akka itself, so you have to provide those as well. Your first Milestone is that within your container you see that the Actor System is exposed from the blended-akka bundle. There is no need to provide another Activator derived from BlendedAkkaActivator. 

* The second dependency is *whenTypeSafeConfigAvailable* which brings an implicit dependency on the [ContainerIdentifierService](https://github.com/woq-blended/blended/blob/master/blended-container-context/src/main/scala/blended/container/context/ContainerIdentifierService.scala). This will determine the *ContainerConfigDir* via the [ContainerContext](https://github.com/woq-blended/blended/blob/master/blended-container-context/src/main/scala/blended/container/context/ContainerContext.scala). If you dan't specify anything else, this will be "./etc". So, the second milestone is to check whether the bundle [blended-container-context](https://github.com/woq-blended/blended/tree/master/blended-container-context) exposes the [ContainerIdentifierService](https://github.com/woq-blended/blended/blob/master/blended-container-context/src/main/scala/blended/container/context/ContainerIdentifierService.scala) instance.

As I said above there are a few dependencies that need to present in your OSGI container. We are currently working on a packaging mechanism and have also specified a list of bundles that form our base container. You can have a look at our [bundle list](https://github.com/woq-blended/blended/blob/master/blended-launcher-features/src/main/resources/blended-base.conf). This list also has our updater bundles in it as we intend all blended containers to benefit from an incremental update mechanism with the next release. You won't need the update bundles for a simple setup, but the rest of list should cover everything you need to get Akka up and running together with Domino activators. 

Once everything is available, from the developers perspective it should be as simple as [in here](https://github.com/woq-blended/blended/blob/master/blended-mgmt-agent/src/main/scala/blended/mgmt/agent/internal/AgentActivator.scala).

Finally, it would be much easier to help you if provided sample code or share the setup of your container or some log files if you have.  

Also - as we are working towards a 2.0 release we would be glad to help you coming to a working Blended-2.0 container based on our build tools if that is an option for you. This would help us to create a useful tutorial for getting started with blended - or even better: contribute one;). 

This having said, for reasons discussed [here](https://github.com/woq-blended/blended/issues/9) we are going to stick with a maven based build process for now. So, if you want to go ahead with SBT, we can try to help, but it is not our primary focus. 


I hope this helps 
Andreas

**Comment by PhilAndrew** [comment]
Thanks, now I got my three SBT projects to work with your code, so this runs:

    whenBundleActive {
      println("runsbt: bundle active in runsbt_control")
      whenActorSystemAvailable { cfg => {
        println("runsbt: actor system available in runsbt_control")
        setupBundleActor(cfg, SomeActor.props(cfg, "blended"))
      }
      }
    }

I have not tried to test an actor. Now I might upload the sbt projects to github.

**Comment by atooni** [comment]
Hi, 

good to hear it works for you. If I understand correctly, you have created 2 projects which do nothing more than exposing the Activators that have been defined in blended-container-context and blended-akka. 

TBH. I do not quite understand why creating those projects was necessary for you. Was it because you wanted something usable with SBT ?  - Within a normal blended project the only module you would have created is the third one with your custom code. 

Capsules can be nested inside each other and I believe what you are asking is can be seen in [ActorSystemCapsule](https://github.com/woq-blended/blended/blob/master/blended-akka/src/main/scala/blended/akka/internal/ActorSystemCapsule.scala) in conjunction with [ActorSystemWatching](https://github.com/woq-blended/blended/blob/master/blended-akka/src/main/scala/blended/akka/ActorSystemWatching.scala). 

In this code we combine 2 capsules and provide then convenience method *whenActorSystemAvailable*. 

We tend to keep the outer capsule to be *whenBundleActive* as this is kind of the direct equivalent of an OSGI Activator without domino involved. 

If you don't mind I ll close the ticket as your question seems to be answered ?

Best regards
Andreas

**Comment by atooni** [comment]
Please let us know once you have published your projects, so that we can share and learn.

**Comment by PhilAndrew** [comment]
In my case I have a few requirements which may be different to yours.
1. Configuration should not come from files in file system, your default is files in Karaf, I'd like to avoid that.
2. Builds must be sbt across all projects as I may not be doing OSGi at all times and most Scala projects are SBT, whether that is painful or not.
3. Only dependency on Karaf is to be the features.xml file and bundle should be mostly normal OSGi as I may have to pull away from Karaf depending on the future.
4. Build and deployment to Karaf will be automated as features file will be automatically placed in Karaf deploy folder and OSGi JAR file pulled from remote URL.
5. Need to have Akka working in OSGi, your code makes that a good thing.

**Comment by atooni** [comment]
I see, In Blended 2.0 we are actually replacing the Karaf runtime by our own launcher und updater framework - all written in Scala. 

I would be interested - where does your config come from if not from the file system ? Are you only packaging it in the jar file ? - What if you have to configure something that is really variable like a database url or a jms connection url ...

**Comment by PhilAndrew** [comment]
Configuration can come from a cloud configuration etcd or ZooKeeper http://sysgears.com/articles/managing-configuration-of-distributed-system-with-apache-zookeeper/. Not implemented yet. Do not want it to be packaged in the jar, all data should be "out-there" in a database or other.
The use case for me is mostly on cloud systems, do not want to be dependent on Karaf.

**Comment by atooni** [comment]
Interesting article - thanks. I will close the ticket for now - let us know how you get on eventually.

**Comment by PhilAndrew** [comment]
Yes I'll put the code online, I'm trying to understand the code around the OSGi and Actors line by line
